toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go

```diff
@@ -658,25 +664,139 @@ func convertPackageVersionToTdnfArg(pkgVer *pkgjson.PackageVer) (tdnfArg string)
+	// 1. Toolchain.
+	// 2. Locally-built packages.
+	// 3. Local cache of packages downloaded from external sources.
+	// 4. Upstream repositories requiring network access.
+	if RepoFlagToolchain&reposFlags != 0 {
+		previousReposList = append(previousReposList, fmt.Sprintf("--enablerepo=%s", repoIDToolchain))
+		r.reposArgsList = append(r.reposArgsList, previousReposList)
+	}
+
+	if RepoFlagLocalBuilds&reposFlags != 0 {
+		previousReposList = append(previousReposList, fmt.Sprintf("--enablerepo=%s", repoIDBuilt))
+		r.reposArgsList = append(r.reposArgsList, previousReposList)
+	}
+
+	if RepoFlagDownloadedCache&reposFlags != 0 {
+		previousReposList = append(previousReposList, fmt.Sprintf("--enablerepo=%s", r.repoIDCache))
+		r.reposArgsList = append(r.reposArgsList, previousReposList)
+	}
+
+	// Options past this point are only valid if upstream repos are enabled.
+	if RepoFlagUpstream&reposFlags == 0 {
+		return
+	}
+
+	previousReposList = append(previousReposList, fmt.Sprintf("--enablerepo=%s", repoIDAll))
+
+	if RepoFlagPreview&reposFlags == 0 {
+		previousReposList = append(previousReposList, fmt.Sprintf("--disablerepo=%s", repoIDPreview))
+	}
+
+	if RepoFlagMarinerDefaults&reposFlags == 0 {
+		previousReposList = append(previousReposList, r.disabledDefaultMarinerReposArgs()...)
+	}
+
+	r.reposArgsList = append(r.reposArgsList, previousReposList)
+}
+
+func (r *RpmRepoCloner) disabledDefaultMarinerReposArgs() (args []string) {
+	args = make([]string, len(r.defaultMarinerRepoIDs))
+	for i, repoID := range r.defaultMarinerRepoIDs {
+		args[i] = fmt.Sprintf("--disablerepo=%s", repoID)
+	}
+
+	return
+}
+
+func (r *RpmRepoCloner) refreshPackagesCache() (err error) {
+	releaseverCliArg, err := tdnf.GetReleaseverCliArg()
+	if err != nil {
+		return
+	}
+
+	args := []string{
+		"makecache",
+		releaseverCliArg,
+		fmt.Sprintf("--enablerepo=%s", repoIDAll),
+	}
+
+	stdout, stderr, err := shell.Execute("tdnf", args...)
+	if err != nil {
+		logger.Log.Errorf("Failed to run 'tdnf makecache'. Stdout:\n%s\nStderr:\n%s\nError: %s.", stdout, stderr, err)
 	}
+
+	return
+}
+
+func readRepoIDs(repoFilePath string) (repoIDs []string, err error) {
+	repoFile, err := os.Open(repoFilePath)
+	if err != nil {
+		return
+	}
+	defer repoFile.Close()
+
+	scanner := bufio.NewScanner(repoFile)
+	for scanner.Scan() {
+		matches := repoIDRegex.FindStringSubmatch(scanner.Text())
+		if len(matches) <= repoIDIndex {
+			continue
+		}
+
+		repoID := matches[repoIDIndex]
+		repoIDs = append(repoIDs, repoID)
+
+		logger.Log.Debugf("Found repo ID: %s", repoID)
+	}
+
+	err = scanner.Err()
+	return
+
+}
+
+func (r *RpmRepoCloner) reposArgsHaveOnlyLocalSources(reposArgs []string) bool {
+	const repoIDIndex = 1
+
+	for _, repoArg := range reposArgs {
+		if strings.Contains(repoArg, "--enablerepo=") {
+			repoID := strings.Split(repoArg, "=")[repoIDIndex]
+			if r.externalRepoIDs[repoID] {
+				return false
```

Comment:
I think we need to explicitly track the toolchain + out/RPMS repos since the user can supply arbitrarily many external repos.

Reply:
Per offline discussion I get the logic now. It may be more clear to explicitly check against the out/RPMs or toolchain repo IDs just for readibility however.

Reply:
Sounds good.